cpp/src/arrow/array/array_nested.cc

```diff
@@ -836,24 +845,32 @@ Result<std::shared_ptr<Array>> MapArray::FromArraysInternal(
 
   using OffsetArrayType = typename TypeTraits<OffsetArrowType>::ArrayType;
   const auto& typed_offsets = checked_cast<const OffsetArrayType&>(*offsets);
-  auto buffers = BufferVector({nullptr, typed_offsets.values()});
+
+  BufferVector buffers;
+  int64_t null_count;
+  if (null_bitmap != nullptr) {
+    buffers = BufferVector({std::move(null_bitmap), typed_offsets.values()});
+    null_count = null_bitmap->size();
```

Comment:
`null_bitmap` was moved-from, so it's invalid to call a method on it after that. It all works fine because the parameter is `const shared_ptr&`, but it should be just `shared_ptr<Buffer>` so you can `std::move` it meaningfully here. This is present in the signature suggestion I've made above.

Another thing: is this logic correct? Is `null_count` the `::size()` of the buffer?

Reply:
> Another thing: is this logic correct? Is `null_count` the `::size()` of the buffer?

Oh, sorry, I totally missed that. That indeed does not look correct

Reply:
`null_count` being the size of the validity bitmap buffer sounded correct but thinking about it now, the size is in bytes and I was thinking it might be sum of bits at that time (1 for each missing/null).

I double checked in Python now and yes, `::size` is wrong. Should set `-1` then? As I am not sure the `null_count` can be retrieved from a `Buffer`.

Reply:
Aha, I can look here and do something similar =)

https://github.com/apache/arrow/blob/524a463207fdb799b2cd784b7ef95052165882ec/cpp/src/arrow/array/data.cc#L210-L211

Reply:
Done here: https://github.com/apache/arrow/pull/41956/commits/043a8ee9ef4e8e2e2a405d27053d83776b3522d0

Reply:
> might be sum of bits at that time (1 for each missing/null

Note that for null_count don't want the sum / count of set bits, but the count of unset bits (as this is a validity_bitmap, although the keyword is confusingly named null_bitmap)

Reply:
Ah, sorry for being sloppy. I have changed back to `-1` as the `null_count` is defined later anyways. Update: https://github.com/apache/arrow/pull/41956/commits/9b4b5c1c72d2fe6ff3a6afeb32511a09440c63e0

Reply:
Keep it as `kUnknownNullCount` and let it be lazily calculated later if someone needs it and call `GetNullCount()` instead of `null_count()`.

Reply:
Done:
https://github.com/apache/arrow/pull/41956/commits/5b6ec5498f3e170d56f64a425dd462a2c8a268e2
https://github.com/apache/arrow/pull/41956/commits/83adabeb1edf62315a95f9adac398177e3d1c9d3